File: ruby/lib/rock_paper_scissors.rb
class Game

  PAIRS = {
    rock:     { beats: :scissors },
    paper:    { beats: :rock },
    scissors: { beats: :paper }
  }.freeze

  # PAIRS = {
  #   rock: {rock: [0,0], paper: [-1,1], scissors: [1,-1], lizard: [1,-1], spock: [-1,1]},
  #   paper: {rock: [1,-1], paper: [0,0], scissors: [-1,1], lizard: [-1,1], spock: [1,-1]},
  #   scissors: {rock: [-1,1], paper: [1,-1], scissors: [0,0], lizard: [1,-1], spock: [-1,1]},
  #   lizard: {rock: [-1,1], paper: [1,-1],scissors: [-1,1],lizard: [0,0], spock: [1,-1]},
  #   spock: {rock: [1,-1], paper: [-1,1], scissors: [1,-1], lizard: [-1,1], spock: [0,0]}

  # }

  def initialize(player1, player2)
    @player1, @player2 = player1, player2
  end

  def winner
    return nil if same_pick?

    if PAIRS[@player1.pick][:beats] == @player2.pick
      @player1
    else
      @player2
    end
  end

  private

  def same_pick?
    @player1.pick == @player2.pick
  end

end

class Player

  attr_reader :pick, :name

  def initialize(name)
    @name = name
  end

  def picks(pick)
    @pick = pick
  end

end

File: javascript/lib/RockPaperScissors.js
function Player(playerName){
	this.playerName = playerName;
}

// Prototype used to create a method //
Player.prototype.picks = function(pick){
   this.pick = pick;
}

// Initialize players. 'this' is the equivalent of the @ symbol//
function Game(player1, player2){
	this.player1 = player1;
	this.player2 = player2;
}

Game.prototype.winner = function(){
	var pairs = {
		rock : {
			beats : 'scissors',
			beats1 : 'lizard'
		},
		paper : {
			beats : 'rock',
			beats1 : 'spock'
		},	
		scissors : {
			beats : 'paper',
			beats1 : 'lizard'
		},
		lizard: {
			beats: 'paper',
			beats1 :'spock'
		},
		spock : {
			beats :'rock',
			beats1 : 'scissors'
		}
	}

	if (this.samePick() == true) {
		return null;
	}

	if(pairs[this.player1.pick]['beats'] == this.player2.pick || pairs[this.player1.pick]['beats1'] == this.player2.pick) {
		this.winnerPick = player1.pick
		this.loserPick = player2.pick
		return player1;
	} else {
		this.winnerPick = player2.pick
		this.loserPick = player1.pick
		return player2;
	}
}

Game.prototype.samePick = function(){
	return this.player1.pick == this.player2.pick ? true : false 
}

Game.prototype.showWinner = function() {
	this.winner();
	var messages = {
		scissorspaper : 'cut',
		paperrock : 'covers',
		rocklizard : 'crushes',
		lizardspock : 'poisons',
		spockscissors : 'smashes',
		scissorslizard : 'decapitate',
		lizardpaper : 'eats',
		paperspock : 'disproves',
		spockrock : 'vaporizes',
		rockscissors : 'breaks'
	}
	
	if(this.winnerPick == undefined) {
		return 'Its a draw!';
	} else {
		messageIndex = this.winnerPick.concat(this.loserPick);
		return this.winnerPick + ' ' + messages[messageIndex]  + ' ' + this.loserPick;
	}

}

